File: programmers/기능개발.js
/*
기능개발
https://programmers.co.kr/learn/courses/30/lessons/42586?language=javascript
*/
function solution(progresses, speeds) {
    let answer = [];
    let queue = [];
    let day = 0;
    for (let i = 0; i < progresses.length; i++) {
        let required = 100 - progresses[i];
        queue.push(Math.ceil(required/speeds[i]));
    }
    
    while (queue.length) {
        console.log(queue)    
        if (day < queue[0]) {
            day += (queue[0] - day);
        }
        else {
            let count = 0;
            while (queue && day >= queue[0]) {
                queue.shift()
                count += 1;
            }
            answer.push(count);
        }
    }
    console.log(answer)
    return answer;
}

File: programmers/기능개발.java
import java.util.ArrayList;
import java.util.Deque;
import java.util.ArrayDeque;
import java.lang.Math;
class Solution {
    public int[] solution(int[] progresses, int[] speeds) {
        ArrayList<Integer> answer = new ArrayList();
        Deque<Double> queue =  new ArrayDeque<Double>();
        int day = 0;
        
        for(int i = 0; i < progresses.length; i++) {
            double required = 100 - progresses[i];
            queue.add(Math.ceil(required/speeds[i]));
        }

        while(!queue.isEmpty()) {
            if (day < queue.getFirst()) {
                day += (queue.getFirst() - day);
            }
            else {
                int count = 0;
                while (!queue.isEmpty() && day >= queue.getFirst()) {
                    queue.removeFirst();
                    count += 1;
                }
                answer.add(count);                
            }   
        }
        return answer.stream().mapToInt(x -> x).toArray();
    }
}